wgengine/filter/filter.go

```diff
@@ -138,15 +138,23 @@ var acceptBucket = rate.NewLimiter(rate.Every(10*time.Second), 3)
 var dropBucket = rate.NewLimiter(rate.Every(5*time.Second), 10)
 
 func (f *Filter) logRateLimit(runflags RunFlags, b []byte, q *packet.QDecode, r Response, why string) {
-	var qs string
-	if q == nil {
-		qs = fmt.Sprintf("(%d bytes)", len(b))
-	} else {
-		qs = q.String()
-	}
+	// Note: it is crucial that q.String() be called only after {accept,drop}Bucket.Allow() passes,
+	// since it causes an allocation.
 	if r == Drop && (runflags&LogDrops) != 0 && dropBucket.Allow() {
+		var qs string
+		if q == nil {
+			qs = fmt.Sprintf("(%d bytes)", len(b))
+		} else {
+			qs = q.String()
+		}
 		f.logf("Drop: %v %v %s\n%s", qs, len(b), why, maybeHexdump(runflags&HexdumpDrops, b))
 	} else if r == Accept && (runflags&LogAccepts) != 0 && acceptBucket.Allow() {
+		var qs string
```

Comment:
This duplication is unfortunate. Can you instead do something like:

```go
      var logVerdict string
      if r == Drop && ... {
              logVerdict = "Drop"
      } else if r == Accept && ... {
               logVerdict = "Accept"
      }
      if logVerdict != "" {
               var qs string   /* .... */
               f.logf("%s: %v %v ...", logVerdict, ....)
      }
```

If that doesn't mess up the escape analysis, that is.